backends/bmv2/simple_switch/midend.cpp

```diff
@@ -119,6 +119,7 @@ SimpleSwitchMidEnd::SimpleSwitchMidEnd(CompilerOptions& options, std::ostream* o
             new P4::CompileTimeOperations(),
             new P4::TableHit(&refMap, &typeMap),
             new P4::EliminateSwitch(&refMap, &typeMap),
+            new P4::TypeChecking(&refMap, &typeMap, true),
```

Comment:
You are fixing a symptom and not the actual problem.
If a pass needs to make sure that types are known and fresh it should run typechecking itself.
The next pass, RemoveLeftSlices does this (but without the 'true' flag).
Perhaps this has to be done just before the backend. The types you set here are not guaranteed to be accurate by the time you reach the backend anyway.
Moreover, you should add a check in the code generating part of the backend to give an error if a type is Type_Unknown instead of generating incorrect code.

Reply:
I moved down TypeChecking as much as possible.